libraries/AP_MSP/AP_MSP_Telem_DJI.cpp

```diff
@@ -108,17 +112,139 @@ MSPCommandResult AP_MSP_Telem_DJI::msp_process_out_esc_sensor_data(sbuf_t *dst)
+    // alternate max distance with traveled distance every 2 seconds
+    if (msp->_msp_status.slow_flashing_on) {
+        home_state.home_distance_m = osd->get_stats_info().max_dist_m;
+    } else {
+        home_state.home_distance_m = osd->get_stats_info().last_distance_m;
+    }
+    home_state.rel_altitude_cm = osd->get_stats_info().max_alt_m * 100;
+#endif
+}
+
+void AP_MSP_Telem_DJI::update_battery_state(battery_state_t &_battery_state)
+{
+    AP_MSP_Telem_Backend::update_battery_state(_battery_state);
+#if OSD_ENABLED
+    if (!osd_stats_item_enabled()) {
+        return;
+    }
+    AP_OSD *osd = AP::osd();
+    if (osd == nullptr) {
+        return;
+    }
+    WITH_SEMAPHORE(osd->get_semaphore());
+    // override telemetry with max current and voltage info
+    _battery_state.batt_current_a = osd->get_stats_info().max_current_a;
+    _battery_state.batt_voltage_v = osd->get_stats_info().min_voltage_v;
+#endif
+}
+
+void AP_MSP_Telem_DJI::update_gps_state(gps_state_t &gps_state)
+{
+    AP_MSP_Telem_Backend::update_gps_state(gps_state);
+#if OSD_ENABLED
+    if (!osd_stats_item_enabled()) {
+        return;
+    }
+    AP_OSD *osd = AP::osd();
+    if (osd == nullptr) {
+        return;
+    }
+    WITH_SEMAPHORE(osd->get_semaphore());
+    // override telemetry with max speed info
+    gps_state.speed_cms = osd->get_stats_info().max_speed_mps * 100;
+#endif
+}
+
+void AP_MSP_Telem_DJI::update_airspeed(airspeed_state_t &airspeed_state)
+{
+    AP_MSP_Telem_Backend::update_airspeed(airspeed_state);
+#if OSD_ENABLED
+    if (!osd_stats_item_enabled()) {
+        return;
+    }
+    AP_OSD *osd = AP::osd();
+    if (osd == nullptr) {
+        return;
+    }
+    WITH_SEMAPHORE(osd->get_semaphore());
+    // override telemetry with max speed info
+    airspeed_state.airspeed_estimate_ms = osd->get_stats_info().max_airspeed_mps;
+#endif
+}
+
+void AP_MSP_Telem_DJI::update_flight_mode_str(char *flight_mode_str, uint8_t size, bool wind_enabled)
+{
+    AP_MSP_Telem_Backend::update_flight_mode_str(flight_mode_str, size, wind_enabled);
+#if OSD_ENABLED
+    if (!osd_stats_item_enabled()) {
+        return;
+    }
+    AP_Stats *stats = AP::stats();
+    if (stats != nullptr) {
+        uint32_t t = stats->get_flight_time_s();
+        // need to check snprintf return value to prevent format-truncation warning in GCC because of -Werror=format-truncation
+        if (snprintf(flight_mode_str, size, "%s %3u:%02u", "STATS", unsigned(t/60), unsigned(t%60)) < 0) {
+            snprintf(flight_mode_str, size, "%s", "STATS --:--");
+        }
+    } else {
+        snprintf(flight_mode_str, size, "%s", "STATS");
+    }
+#endif
+}
+
+float AP_MSP_Telem_DJI::get_rssi() const
+{
+    float rssi = AP_MSP_Telem_Backend::get_rssi();
+#if OSD_ENABLED
+    if (!osd_stats_item_enabled()) {
+        return rssi;
+    }
+    AP_RSSI* ap_rssi = AP::rssi();
+    if (ap_rssi == nullptr) {
+        return 0;
```

Comment:
I guess this is kind of an error return value?  It's not distinctive...

Reply:
> There seems to be more flashing when in stats mode than should be including "ACRO" which makes no sense...
not sure I understand this?
Flashing right now is orthogonal to stats mode,  it would flash irregardless, are you implying that when stats are enabled we should suspend any kind of flashing?
> and it is claiming 14 satellites when the plane hasn't seen a single one.
14 sats when no fix is a known DJI bug, I never really tried to find a hack for it